File: lib/sidereon/gnss/velocity.ex
defmodule Sidereon.GNSS.Velocity do
  @moduledoc """
  Recover receiver velocity and clock drift from one epoch of range-rate or
  Doppler observations against a precise SP3 or broadcast ephemeris source.

  The numerical model and least-squares solve live in the Rust GNSS core. This
  module preserves the Elixir API shape: input normalization, per-satellite
  option resolution, and public result/error maps.
  """

  alias Sidereon.GNSS.{Broadcast, SP3, Time}
  alias Sidereon.GNSS.Core.Constants
  alias Sidereon.GNSS.Core.Types
  alias Sidereon.NIF

  @type vec3 :: {float(), float(), float()}
  @type receiver :: vec3() | %{x_m: number(), y_m: number(), z_m: number()}
  @type observation :: {String.t(), number()}

  @type result :: %{
          velocity_m_s: vec3(),
          speed_m_s: float(),
          clock_drift_s_s: float(),
          residuals_m_s: %{String.t() => float()},
          used_sats: [String.t()],
          n_satellites: non_neg_integer()
        }

  @doc """
  Solve for receiver velocity and clock drift at one receive epoch.

  `observations` are `{satellite_id, value}` pairs. Values are pseudorange rates
  in m/s by default, or Doppler shifts in Hz with `observable: :doppler`.
  """
  @spec solve(SP3.t() | Broadcast.t(), [observation()], NaiveDateTime.t(), receiver(), keyword()) ::
          {:ok, result()} | {:error, term()}
  def solve(source, observations, epoch, receiver_position, opts \\ [])

  def solve(%SP3{} = source, observations, %NaiveDateTime{} = epoch, receiver_position, opts)
      when is_list(observations) do
    do_solve(source, observations, epoch, receiver_position, opts)
  end

  def solve(%Broadcast{} = source, observations, %NaiveDateTime{} = epoch, receiver_position, opts)
      when is_list(observations) do
    do_solve(source, observations, epoch, receiver_position, opts)
  end

  def solve(%SP3{}, observations, %NaiveDateTime{}, _receiver, _opts) when not is_list(observations),
    do: {:error, :no_observations}

  def solve(%Broadcast{}, observations, %NaiveDateTime{}, _receiver, _opts) when not is_list(observations),
    do: {:error, :no_observations}

  defp do_solve(source, observations, epoch, receiver_position, opts) do
    observable = Keyword.get(opts, :observable, :range_rate)
    carrier_hz = Keyword.get(opts, :carrier_hz, Constants.gps_l1_hz()) * 1.0
    carrier_fun = carrier_hz_fun(Keyword.get(opts, :carrier_hz_by_sat), carrier_hz)
    sat_drift_fun = sat_clock_drift_fun(Keyword.get(opts, :sat_clock_drift))
    light_time? = Keyword.get(opts, :light_time, true)
    sagnac? = Keyword.get(opts, :sagnac, true)

    with :ok <- ensure_nonempty(observations),
         {:ok, receiver} <- Types.normalize_ecef(receiver_position),
         {:ok, normalized} <- normalize_observations(observations, observable, carrier_fun),
         terms = observation_terms(normalized, sat_drift_fun),
         {:ok, core_result} <-
           core_solve(source, terms, epoch, receiver, observable, light_time?, sagnac?) do
      {:ok, to_result_map(core_result)}
    end
  end

  @doc """
  Convert a Doppler shift in Hz to a pseudorange rate in m/s.
  """
  @spec doppler_to_range_rate(number(), number()) :: float()
  def doppler_to_range_rate(doppler_hz, carrier_hz \\ Constants.gps_l1_hz()) do
    NIF.velocity_doppler_to_range_rate(doppler_hz * 1.0, carrier_hz * 1.0)
  end

  @doc """
  Convert a pseudorange rate in m/s to a Doppler shift in Hz.
  """
  @spec range_rate_to_doppler(number(), number()) :: float()
  def range_rate_to_doppler(rho_dot_m_s, carrier_hz \\ Constants.gps_l1_hz()) do
    NIF.velocity_range_rate_to_doppler(rho_dot_m_s * 1.0, carrier_hz * 1.0)
  end

  defp ensure_nonempty([]), do: {:error, :no_observations}
  defp ensure_nonempty(_), do: :ok

  defp normalize_observations(observations, observable, carrier_fun) do
    Enum.reduce_while(observations, {:ok, [], MapSet.new()}, fn entry, {:ok, acc, seen} ->
      case normalize_one(entry, observable, carrier_fun) do
        {:ok, {sat, _system_letter, _prn, _value, _carrier} = normalized} ->
          if MapSet.member?(seen, sat) do
            {:halt, {:error, {:duplicate_observation, sat}}}
          else
            {:cont, {:ok, [normalized | acc], MapSet.put(seen, sat)}}
          end

        {:error, _} = err ->
          {:halt, err}
      end
    end)
    |> case do
      {:ok, acc, _seen} -> {:ok, Enum.reverse(acc)}
      {:error, _} = err -> err
    end
  end

  defp normalize_one({sat, value}, :range_rate, _carrier_fun) when is_binary(sat) and is_number(value) do
    with {:ok, system_letter, prn} <- Types.parse_sat_id(sat) do
      {:ok, {sat, system_letter, prn, value * 1.0, Constants.gps_l1_hz()}}
    end
  end

  defp normalize_one({sat, value}, :doppler, carrier_fun) when is_binary(sat) and is_number(value) do
    with {:ok, system_letter, prn} <- Types.parse_sat_id(sat),
         {:ok, carrier} <- carrier_fun.(sat) do
      {:ok, {sat, system_letter, prn, value * 1.0, carrier}}
    end
  end

  defp normalize_one(entry, _observable, _carrier_fun), do: {:error, {:invalid_observation, entry}}

  defp observation_terms(normalized, sat_drift_fun) do
    Enum.map(normalized, fn {sat, system_letter, prn, value, carrier_hz} ->
      {system_letter, prn, value, carrier_hz, sat_drift_fun.(sat)}
    end)
  end

  defp core_solve(_source, [], _epoch, _receiver, _observable, _light_time?, _sagnac?),
    do: {:error, {:too_few_satellites, 0, 4}}

  defp core_solve(%SP3{handle: handle}, terms, epoch, receiver, observable, light_time?, sagnac?) do
    {jd_whole, jd_fraction} = Time.epoch_to_split_jd(epoch)

    case NIF.sp3_velocity_solve(
           handle,
           terms,
           jd_whole,
           jd_fraction,
           receiver,
           Atom.to_string(observable),
           light_time?,
           sagnac?
         ) do
      {:ok, result} -> {:ok, result}
      {:error, _} = err -> err
      other -> {:error, other}
    end
  rescue
    e in ErlangError -> {:error, e.original}
  end

  defp core_solve(%Broadcast{handle: handle}, terms, epoch, receiver, observable, light_time?, sagnac?) do
    with {:ok, t_j2000_s} <- Time.epoch_to_j2000_seconds_fractional(epoch) do
      case NIF.broadcast_velocity_solve(
             handle,
             terms,
             t_j2000_s,
             receiver,
             Atom.to_string(observable),
             light_time?,
             sagnac?
           ) do
        {:ok, result} -> {:ok, result}
        {:error, _} = err -> err
        other -> {:error, other}
      end
    end
  rescue
    e in ErlangError -> {:error, e.original}
  end

  defp to_result_map({velocity, speed, clock_drift, residuals, used_sats}) do
    %{
      velocity_m_s: velocity,
      speed_m_s: speed,
      clock_drift_s_s: clock_drift,
      residuals_m_s: Map.new(residuals),
      used_sats: used_sats,
      n_satellites: length(used_sats)
    }
  end

  defp carrier_hz_fun(nil, default_hz) do
    fn sat -> normalize_carrier(default_hz, sat) end
  end

  defp carrier_hz_fun(map, _default_hz) when is_map(map) do
    fn sat ->
      case Map.fetch(map, sat) do
        {:ok, nil} -> {:error, {:missing_carrier, sat}}
        {:ok, hz} -> normalize_carrier(hz, sat)
        :error -> {:error, {:missing_carrier, sat}}
      end
    end
  end

  defp carrier_hz_fun(fun, _default_hz) when is_function(fun, 1) do
    fn sat ->
      case fun.(sat) do
        nil -> {:error, {:missing_carrier, sat}}
        hz -> normalize_carrier(hz, sat)
      end
    end
  end

  defp normalize_carrier(carrier, _sat) when is_number(carrier) and carrier > 0, do: {:ok, carrier * 1.0}

  defp normalize_carrier(_carrier, sat), do: {:error, {:invalid_carrier, sat}}

  defp sat_clock_drift_fun(nil), do: fn _sat -> 0.0 end

  defp sat_clock_drift_fun(map) when is_map(map), do: fn sat -> (map[sat] || 0.0) * 1.0 end

  defp sat_clock_drift_fun(fun) when is_function(fun, 1), do: fn sat -> fun.(sat) * 1.0 end
end

File: native/sidereon_nif/src/velocity.rs
//! Rustler boundary for GNSS receiver velocity solves.
//!
//! This module decodes loaded ephemeris handles, observation terms, epoch
//! scalars, and receiver position, then calls `sidereon_core::velocity`.
//! No least-squares row construction or Doppler algebra lives at this boundary.

use rustler::{Encoder, Env, NifResult, ResourceArc, Term};
use sidereon_core::observables::j2000_seconds_from_split;
use sidereon_core::velocity::{
    doppler_to_range_rate, range_rate_to_doppler, solve, VelocityError, VelocityObservable,
    VelocityObservation, VelocitySolution, VelocitySolveOptions,
};
use sidereon_core::{GnssSatelliteId, GnssSystem};

use crate::broadcast::BroadcastResource;
use crate::sp3::Sp3Resource;

type Vec3 = (f64, f64, f64);
type ObservationTerm = (String, u8, f64, f64, f64);

mod atoms {
    rustler::atoms! {
        ok,
        error,
        no_observations,
        too_few_satellites,
        singular_geometry,
        duplicate_observation,
        invalid_carrier,
        invalid_observable,
        invalid_epoch,
        invalid_input,
        invalid_observation,
        invalid_receiver_state
    }
}

#[rustler::nif]
pub fn velocity_doppler_to_range_rate(doppler_hz: f64, carrier_hz: f64) -> NifResult<f64> {
    doppler_to_range_rate(doppler_hz, carrier_hz).map_err(crate::errors::invalid_input)
}

#[rustler::nif]
pub fn velocity_range_rate_to_doppler(range_rate_m_s: f64, carrier_hz: f64) -> NifResult<f64> {
    range_rate_to_doppler(range_rate_m_s, carrier_hz).map_err(crate::errors::invalid_input)
}

#[rustler::nif(schedule = "DirtyCpu")]
#[allow(clippy::too_many_arguments)]
pub fn sp3_velocity_solve<'a>(
    env: Env<'a>,
    handle: ResourceArc<Sp3Resource>,
    observations: Vec<ObservationTerm>,
    jd_whole: f64,
    jd_fraction: f64,
    receiver_ecef_m: Vec3,
    observable: String,
    light_time: bool,
    sagnac: bool,
) -> Term<'a> {
    let t_rx_j2000_s = match j2000_seconds_from_split(jd_whole, jd_fraction) {
        Ok(t) => t,
        Err(_) => return (atoms::error(), atoms::invalid_epoch()).encode(env),
    };
    let result = decode_observable(&observable).map(|observable| {
        solve(
            &handle.sp3,
            &decode_observations(observations),
            vec3_to_array(receiver_ecef_m),
            t_rx_j2000_s,
            VelocitySolveOptions {
                observable,
                light_time,
                sagnac,
            },
        )
    });
    encode_nested_result(env, result)
}

#[rustler::nif(schedule = "DirtyCpu")]
#[allow(clippy::too_many_arguments)]
pub fn broadcast_velocity_solve<'a>(
    env: Env<'a>,
    handle: ResourceArc<BroadcastResource>,
    observations: Vec<ObservationTerm>,
    t_rx_j2000_s: f64,
    receiver_ecef_m: Vec3,
    observable: String,
    light_time: bool,
    sagnac: bool,
) -> Term<'a> {
    let result = decode_observable(&observable).map(|observable| {
        solve(
            &handle.store,
            &decode_observations(observations),
            vec3_to_array(receiver_ecef_m),
            t_rx_j2000_s,
            VelocitySolveOptions {
                observable,
                light_time,
                sagnac,
            },
        )
    });
    encode_nested_result(env, result)
}

fn decode_observable(value: &str) -> Result<VelocityObservable, ()> {
    match value {
        "range_rate" => Ok(VelocityObservable::RangeRate),
        "doppler" => Ok(VelocityObservable::Doppler),
        _ => Err(()),
    }
}

fn decode_observations(observations: Vec<ObservationTerm>) -> Vec<VelocityObservation> {
    observations
        .into_iter()
        .filter_map(
            |(system_letter, prn, value, carrier_hz, sat_clock_drift_s_s)| {
                sat_from_parts(&system_letter, prn).map(|satellite_id| VelocityObservation {
                    satellite_id,
                    value,
                    carrier_hz,
                    sat_clock_drift_s_s,
                })
            },
        )
        .collect()
}

fn sat_from_parts(system_letter: &str, prn: u8) -> Option<GnssSatelliteId> {
    let letter = system_letter.chars().next()?;
    let system = GnssSystem::from_letter(letter)?;
    GnssSatelliteId::new(system, prn).ok()
}

fn vec3_to_array(vec: Vec3) -> [f64; 3] {
    [vec.0, vec.1, vec.2]
}

fn array_to_vec3(array: [f64; 3]) -> Vec3 {
    (array[0], array[1], array[2])
}

fn encode_nested_result<'a>(
    env: Env<'a>,
    result: Result<Result<VelocitySolution, VelocityError>, ()>,
) -> Term<'a> {
    match result {
        Ok(Ok(solution)) => encode_solution(env, &solution),
        Ok(Err(error)) => encode_error(env, error),
        Err(()) => (atoms::error(), atoms::invalid_observable()).encode(env),
    }
}

fn encode_solution<'a>(env: Env<'a>, solution: &VelocitySolution) -> Term<'a> {
    let residuals: Vec<(String, f64)> = solution
        .residuals_m_s
        .iter()
        .map(|(sat, residual)| (sat.to_string(), *residual))
        .collect();
    let used_sats: Vec<String> = solution.used_sats.iter().map(ToString::to_string).collect();
    (
        atoms::ok(),
        (
            array_to_vec3(solution.velocity_m_s),
            solution.speed_m_s,
            solution.clock_drift_s_s,
            residuals,
            used_sats,
        ),
    )
        .encode(env)
}

fn encode_error<'a>(env: Env<'a>, error: VelocityError) -> Term<'a> {
    let reason = match error {
        VelocityError::NoObservations => atoms::no_observations().encode(env),
        VelocityError::TooFewSatellites { used, required } => {
            (atoms::too_few_satellites(), used as u64, required as u64).encode(env)
        }
        VelocityError::SingularGeometry => atoms::singular_geometry().encode(env),
        VelocityError::DuplicateObservation { satellite_id } => {
            (atoms::duplicate_observation(), satellite_id.to_string()).encode(env)
        }
        VelocityError::InvalidCarrier { satellite_id } => {
            (atoms::invalid_carrier(), satellite_id.to_string()).encode(env)
        }
        VelocityError::InvalidInput { .. } => atoms::invalid_input().encode(env),
        VelocityError::InvalidObservation { satellite_id } => {
            (atoms::invalid_observation(), satellite_id.to_string()).encode(env)
        }
        VelocityError::InvalidReceiverState => atoms::invalid_receiver_state().encode(env),
    };
    (atoms::error(), reason).encode(env)
}
